feat(proxy): version 1

julien-f opened the pull request:
Fixes #4254 

## Deploy proxy

### Click on proxies

![image](https://user-images.githubusercontent.com/21563339/71256465-17546d00-2331-11ea-89f6-de63b3ddb42b.png)

### Click on "Deploy a proxy"

![deployProxyButton](https://user-images.githubusercontent.com/21563339/80114306-b6110480-8583-11ea-8722-83f22e5be778.png)

### Select destination SR

![image](https://user-images.githubusercontent.com/21563339/80114537-fe302700-8583-11ea-9bf5-598e0b143021.png)

### Select destination network

![image](https://user-images.githubusercontent.com/21563339/80114619-11db8d80-8584-11ea-8de2-9ac544556061.png)

### Set static IP

![image](https://user-images.githubusercontent.com/21563339/80114712-2ae43e80-8584-11ea-828d-92e1b3bc91f0.png)

### Deployed proxies

![image](https://user-images.githubusercontent.com/21563339/80114941-71d23400-8584-11ea-8559-9892cf929f29.png)

## Create proxy remote

### Select a proxy in the remote form

![image](https://user-images.githubusercontent.com/21563339/80117316-54529980-8587-11ea-9721-8a1d61293efe.png)

### Proxy remotes

![image](https://user-images.githubusercontent.com/21563339/80117185-23726480-8587-11ea-8136-50a8d98b4a5a.png)


## Create proxy backup

### Select a proxy in the backup form

![image](https://user-images.githubusercontent.com/21563339/80116365-29b41100-8586-11ea-9746-e01ca3e53996.png)

### If simple backup, select the linked remote proxy

![image](https://user-images.githubusercontent.com/21563339/80116441-44868580-8586-11ea-9177-ab150669e78a.png)

### Proxy backups

![image](https://user-images.githubusercontent.com/21563339/80116831-bb238300-8586-11ea-8979-1a4e96132837.png)

### Check list

> Check items when done or if not relevant

- [x] PR reference the relevant issue (e.g. `Fixes #007`)
- [ ] if UI changes, a screenshot has been added to the PR
- [ ] `CHANGELOG.unreleased.md`:
   - enhancement/bug fix entry added
   - list of packages to release updated (`${name} v${new version}`)
- [ ] documentation updated
- [ ] **I have tested added/updated features** (and impacted code)

### Process

1. create a PR as soon as possible
1. mark it as `WiP:` (Work in Progress) if not ready to be merged
1. when you want a review, add a reviewer (and only one)
1. if necessary, update your PR, and re- add a reviewer

julien-f committed:
feat(proxy): disaster recovery

julien-f committed:
feat(proxy/api): disable timeout for incoming requests

julien-f committed:
fix(CHANGELOG): revert changes

julien-f commented on packages/xo-server/src/xo-mixins/backups-ng/index.js:
```diff
@@ -91,6 +91,7 @@ export type BackupJob = {|
   ...$Exact<Job>,
   compression?: 'native' | 'zstd' | '',
   mode: Mode,
+  proxy?: 'string',
```
@badrAZ typing is wrong here.

badrAZ replied:
I'll fix it in the @pdonias PR :+1:

julien-f replied:
I'll do it directly on this PR.

julien-f commented on packages/xo-web/src/common/editable/index.js:
```diff
@@ -426,18 +427,19 @@ Select.defaultProps = { multi: false }
   tag: SelectTag,
   vgpuType: SelectVgpuType,
   VM: SelectVm,
-  'VM-template': SelectVmTemplate,
```
@badrAZ Don't introduce unrelated changes.

julien-f committed:
feat(proxy/cli): colored output

julien-f committed:
Merge branch 'master' into proxy-v1

julien-f committed:
chore(xo-web/editable): remove unrelated changes

julien-f committed:
Merge branch 'master' into proxy-v1

julien-f committed:
feat(proxy/api.md): various updates

julien-f committed:
fix(proxy): remove unused better-sqlite3

julien-f committed:
Merge branch 'master' into proxy-v1

julien-f committed:
Merge branch 'master' into proxy-v1

julien-f committed:
chore(proxy): update xo-common

julien-f committed:
feat(proxy/api): better debug log

julien-f committed:
chore(proxy): remove unused deps

julien-f committed:
chore(xapi): update xen-api peer dep

julien-f committed:
chore(proxy): `julien-f-source-map-support` → `source-map-support`

`source-map-support` is correctly maintained and support modern Node versions.

julien-f committed:
chore(xapi/isVmRunning): use `in` operator

julien-f committed:
Merge branch 'master' into proxy-v1

julien-f committed:
feat(xen-api): 0.28.0

julien-f committed:
feat(@xen-orchestra/backups): 0.1.0

julien-f committed:
feat(xapi): requires Node 8.10

julien-f committed:
feat(xapi): publish on `npm version`

julien-f committed:
chore(xapi): remove unused dep

julien-f committed:
chore(xapi): use work email address

julien-f committed:
fix(xapi): `cross` → `cross-env`

julien-f committed:
feat(@xen-orchestra/proxy): 0.1.0

julien-f committed:
fix(proxy/backup.run): should not return an array

julien-f committed:
Merge branch 'master' into proxy-v1

julien-f committed:
feat(@xen-orchestra/proxy): 0.1.1

julien-f committed:
fix(proxy/server): dont delay require to stopping phase

julien-f committed:
chore(proxy/backups): split code into modules

julien-f committed:
feat(proxy): delta backup & continuous replication

julien-f committed:
Merge branch 'master' into proxy-v1

julien-f committed:
feat(proxy/appliance): use support tunnel systemd service

julien-f committed:
feat(proxy/appliance): updater.getLocalManifest

julien-f committed:
feat(proxy/appliance): updater.upgrade

julien-f committed:
feat(proxy): backups.listVmBackups (#4813)

julien-f committed:
fix(xapi): ignore errors when augmenting errors

julien-f committed:
feat(proxy/cli): display complete objects

julien-f committed:
feat(proxy): backup.importVmBackup (#4820)

julien-f committed:
feat(@xen-orchestra/proxy): 0.2.0

julien-f committed:
fix(proxy/appliance.updater.upgrade): update is a notification not a request

julien-f committed:
Merge branch 'master' into proxy-v1

julien-f committed:
Merge branch 'master' into proxy-v1

julien-f committed:
Merge branch 'master' into proxy-v1

julien-f committed:
feat(proxy-cli): move out of proxy pkg

julien-f committed:
chore(proxy): remove unused dev deps

julien-f committed:
feat(@xen-orchestra/proxy-cli): 0.1.0

julien-f committed:
fix(xo-proxy-cli): dont fail on missing config

julien-f committed:
feat(@xen-orchestra/proxy-cli): 0.1.1

julien-f committed:
feat(proxy/api/appliance): implement updater.getState method

julien-f committed:
feat(proxy-cli): --raw flag

julien-f committed:
feat(proxy/api/appliance): implement check method

julien-f committed:
feat(proxy): replace selfsigned by openssl

This module appears to be no longer maintained and is not compatible with Node 12.

See jfromaniello/selfsigned#33

julien-f committed:
fix(proxy): add missing dep getopts

It was incorrectly removed in cc4adef78

julien-f committed:
fix(proxy): await genSelfSignedCert

Introduced by 5d40d5e7d

julien-f committed:
fix(proxy): dont stop whole process on cert/key issue

julien-f committed:
fix(proxy): dont overwrite existing cert/key files

julien-f committed:
Merge branch 'master' into proxy-v1

julien-f committed:
feat(proxy-cli): read response as stream

Dont wait for all response before parsing and displaying.

julien-f committed:
feat(proxy/backup.run): can stream execution logs

julien-f committed:
Merge branch 'master' into proxy-v1

julien-f committed:
Merge branch 'master' into proxy-v1

julien-f committed:
Merge branch 'master' into proxy-v1

julien-f committed:
feat(proxy): remove blank lines around help message

julien-f committed:
feat(proxy/help): display pkg name and version

julien-f committed:
Merge branch 'master' into proxy-v1

julien-f committed:
chore(proxy-cli): use @vates/read-chunk

julien-f committed:
Merge branch 'master' into proxy-v1

julien-f committed:
chore(proxy): use @xen-orchestra/self-signed

julien-f committed:
Merge branch 'master' into proxy-v1

julien-f committed:
chore(proxy/_decorateWith): remove unused module

julien-f committed:
chore(proxy/_parseDuration): remove unused module

julien-f committed:
fix(proxy): failing to change user/group is fatal

julien-f committed:
fix(proxy): correctly declare help and safe-mode opts

julien-f committed:
Merge branch 'master' into proxy-v1

julien-f committed:
chore: normalize packages

julien-f committed:
feat(@xen-orchestra/xapi): 0.1.0

julien-f committed:
feat(@xen-orchestra/backups): 0.2.0

julien-f committed:
feat(@xen-orchestra/proxy): 0.3.0

julien-f committed:
fix(xo-server): revert changes for time being

julien-f committed:
feat(xapi/VM_snapshot): dont set is_a_template = false

julien-f committed:
feat(proxy): increase stack trace limit to 20

julien-f committed:
feat(proxy): ignore stack trace enhancement errors

julien-f committed:
fix(proxy): source-map-support import

Introduced by 95a84aad8

julien-f committed:
feat(proxy): regenerate too small cert

julien-f committed:
Merge branch 'master' into proxy-v1

julien-f committed:
Merge branch 'master' into proxy-v1

julien-f committed:
feat(@xen-orchestra/proxy): 0.4.0

julien-f committed:
fix(proxy/backup.listVmBackups): pass vmUuid instead of dir to listVmBackups

julien-f committed:
fix(proxy/backup.listVmBackups): inject id prop used for import

julien-f committed:
fix(proxy/backup.importVmBackup): invalid RemodeAdapter due to missing await

julien-f committed:
fix(proxy/backup.importVmBackup): pass whole SR record to importDeltaVm

julien-f committed:
fix(proxy/backup.importVmBackup): dont try to find a base VM

julien-f committed:
fix(proxy/backup): log non-enumerable error props

julien-f committed:
feat(@xen-orchestra/proxy): 0.4.1

julien-f committed:
fix(proxy/SyncThenable): correctly store value

julien-f committed:
fix(proxy/serializeError): only handle instances of Error

julien-f committed:
Merge branch 'master' into proxy-v1

julien-f committed:
feat(xen-api): 0.29.0

julien-f committed:
chore(proxy): remove unused dep xen-api

julien-f committed:
fix(xapi/_assertHealthyVdiChain): ignore missing parents

julien-f committed:
feat(proxy/backup.run): more logs

julien-f committed:
feat(@xen-orchestra/xapi): 0.1.1

julien-f committed:
feat(@xen-orchestra/proxy): 0.4.2

julien-f committed:
Revert "chore(proxy): remove unused dep xen-api"

This reverts commit 8653bd5b032fed815540fb9e4fd3aac9da2151f3.

julien-f committed:
feat(@xen-orchestra/proxy): 0.4.3

julien-f committed:
feat(proxy/backup.run): log VMs' id before run

julien-f committed:
feat(proxy/api.md): document backup.run streamLogs param

julien-f committed:
fix(proxy/backup.run): dont run multiple times the same job

julien-f committed:
feat(proxy/backup.run): fail if job already running

julien-f committed:
feat(proxy/Task): easier API based on node-zone

julien-f committed:
fix(proxy-cli): port not mandatory in config

julien-f committed:
feat(proxy/backup.listRunningJobs): new API method

julien-f committed:
fix(proxy/_deltaVm): cancel VDI exports/imports in case of failure

julien-f committed:
feat(@xen-orchestra/xapi): 0.2.0

julien-f committed:
feat(@xen-orchestra/proxy): 0.4.4

julien-f committed:
feat(@xen-orchestra/proxy-cli): 0.1.2

julien-f committed:
fix(proxy/api): send data regularly to avoid connection loss

julien-f committed:
chore(proxy/_VmBackups): create writers in constructors

julien-f committed:
fix(proxy/_importDeltaVm): VDI_create expects `SR` param, not `sr`

This lead to the creation of the VDI on the default SR instead of the configured one.

julien-f committed:
chore(proxy/_exportDeltaVm): remove unused opts

julien-f committed:
chore(proxy/_importDeltaVm): make fullVdisRequired opt a Set of UUIDs

julien-f committed:
feat(@xen-orchestra/proxy): 0.4.5

julien-f committed:
fix(proxy/backups/CR): fix tag

julien-f committed:
Merge branch 'master' into proxy-v1

julien-f committed:
feat(xapi/VM_create): support for setting bios_strings

julien-f committed:
fix(proxy/backups): log all VM level errors

julien-f committed:
feat(proxy/forkStreamUnpipe): add forks prop to source stream

This prop contains the number of active forks for this stream.

julien-f committed:
fix(proxy/full backup & DR): correctly create task

Introduced by 20079bbf5

julien-f committed:
feat(proxy/DeltaBackupWriter): explicit error on missing parent

julien-f committed:
feat(@xen-orchestra/proxy): 0.4.6

julien-f committed:
chore(xapi/_assertHealthyVdiChain): merge child and parent fetching

julien-f committed:
fix(xapi/_assertHealthyVdiChain): dont fail on missing VDIs

julien-f committed:
fix(xapi/_assertHealthyVdiChain): dont fail on missing VDIs (2)

julien-f committed:
feat(@xen-orchestra/xapi): 0.2.1

julien-f committed:
Merge branch 'master' into proxy-v1

julien-f committed:
feat(@xen-orchestra/proxy): 0.4.7

julien-f committed:
Merge branch 'master' into proxy-v1

julien-f committed:
chore(proxy/exportDeltaVm): simpler VDI export syntax

julien-f committed:
feat(proxy/backups): support checkpointSnapshot setting

julien-f committed:
feat(@xen-orchestra/xapi): 0.3.0

julien-f committed:
feat(@xen-orchestra/proxy): 0.5.0

badrAZ committed:
fix(@xen-orchestra/proxy#readDeltaVmBackup): use path.join (#5338)

`path.resolve` will use `process.cwd()` as base if first path is relative, which is not what's wanted here.

julien-f committed:
Merge branch 'master' into proxy-v1

julien-f committed:
feat(@xen-orchestra/proxy): 0.5.1

julien-f committed:
feat(proxy/backup.listVmBackups): remove unused properties

Fixes xoa-support#3093

julien-f committed:
feat(@xen-orchestra/backups): 0.3.0

julien-f committed:
feat(@xen-orchestra/proxy): 0.6.0

badrAZ committed:
chore(proxy): resource management implementation (#5368)

badrAZ committed:
feat(proxy): ability to list disk partitions (#5369)

julien-f committed:
Merge branch 'master' into proxy-v1

badrAZ committed:
fix: yarn.lock (#5396)

badrAZ committed:
feat(proxy): ability to mount/unmount a partition (#5380)

badrAZ committed:
feat(@xen-orchestra/proxy): ability to list partition files (#5394)

julien-f committed:
fix(proxy): backup id should not contain remote id

Fixes xoa-support#3122

badrAZ committed:
feat(proxy,proxy-cli): support binary call result (#5398)

badrAZ committed:
feat(proxy): backup.fetchPartitionFiles (#5395)

julien-f committed:
fix(proxy/mixins): set content-type to application/json on errors

badrAZ committed:
fix(proxy#listPartitions): make partion.type a number (#5402)

julien-f committed:
fix(proxy/api): dont use binary mode for object streams

julien-f committed:
fix(api/importDeltaVm): suspend_VDI handling

badrAZ committed:
fix(package.json): fix "@vates/multi-key-map" version (#5408)

badrAZ committed:
fix(proxy/deduped): invalidate cache on failed acquisition (#5413)

julien-f committed:
Merge branch 'master' into proxy-v1

julien-f committed:
feat(@xen-orchestra/backups): 0.4.0

julien-f committed:
feat(@xen-orchestra/proxy): 0.7.0

badrAZ committed:
fix(proxy/remotes.getHandler): always use configuration (#5419)

badrAZ committed:
feat(proxy, backups): use remote adapters as resources (#5403)

badrAZ committed:
feat(proxy/backups): ability to delete VM backup (#5429)

badrAZ committed:
feat(proxy/backups): ability to list LVM partitions (#5361)

badrAZ committed:
feat(proxy/backups): ability to mount LVM partition (#5424)

badrAZ committed:
feat(proxy/backups): restore file from raw disk (#5440)

badrAZ committed:
feat(proxy/backups): restore file from raw disk with LVM (#5441)

Fixes #5439

badrAZ committed:
chore(proxy/remotes): add FIXME (#5443)

badrAZ committed:
chore(proxy/backups): update comment (#5444)

badrAZ committed:
fix(proxy#listPartition): support recent version of partx

julien-f commented on @xen-orchestra/proxy/src/app/mixins/backups/_listPartitions.js:
```diff
@@ -28,12 +28,19 @@ const parsePartxLine = createParser({
+  }
 
   return parts
     .split(/\r?\n/)
```
Move these transforms in the `try` block and you won't need to declare `parts` as mutable.

badrAZ replied:
What if `parsePartxLine` throws an error?

julien-f replied:
You're right, that makes sense, maybe using `.catch()` to returns an empty string would be a better solution?

julien-f commented on @xen-orchestra/proxy/src/app/mixins/backups/_listPartitions.js:
```diff
@@ -28,12 +28,19 @@ const parsePartxLine = createParser({
+      devicePath,
+    ])
+  } catch (error) {
+    // Partx results with code 1 which throws an error since v2.33
```
```js
    // partx returns 1 since v2.33 when failing to read partitions.
    //
    // Prior versions are correctly handled by the nominal case.
```

badrAZ committed:
fix(proxy/backups#listPartitions): compat with partx >=2.33 (#5445)

badrAZ committed:
feat(proxy/backups): use XAPI connection as resource (#5442)

badrAZ committed:
feat(proxy/backups): block backups in case of invalid license (#5452)

badrAZ committed:
feat(proxy): use the new resource management API (#5458)

Also fix a bug where backups without replications were no longer working introduced in 9148510c0

julien-f committed:
fix(proxy): use promise-toolbox@0.16

Required by b8a3eb624

julien-f committed:
fix(proxy/backup.run): no need running updater in dev

Introduced in 9fddc633a

julien-f committed:
feat(proxy/Backup#run): fetch SRs and remotes in parallel

julien-f committed:
fix(proxy/importDeltaVm): other suspend_VDI fixes

julien-f committed:
feat(xapi/VM_import): support onVmCreation param

julien-f committed:
feat(@xen-orchestra/xapi): 0.4.0

julien-f committed:
fix(proxy/{CR,DR}): add tags at VM creation

Same as 8dcff63ae but for new backup code.

Fixes https://xcp-ng.org/forum/topic/3285/cr-tag-ignored-kind-of-in-smart-backup

Necessary to avoid smart mode from matching the replicated VMs currently being imported.

badrAZ committed:
feat(proxy/backups): stricter perms for backup dirs (#5461)

Same as ae2a92d22 but for new backup code.

julien-f committed:
Merge branch 'master' into proxy-v1

julien-f committed:
chore(proxy): use latest XO packages

julien-f committed:
fix(proxy/backup): correctly set deltaChainLength

julien-f committed:
fix(xapi/VM_getDisks): correctly fetch VBDs

julien-f committed:
feat(proxy): detect missing bases for delta backups and CR

julien-f committed:
feat(@xen-orchestra/proxy): 0.8.0

julien-f committed:
chore(proxy): introduce simple asyncMap

This simplifies the pattern `Promise.all` + `Array#map` but also handles
all iterables and does not trigger `array-callback-return` error from
ESLint.

julien-f committed:
chore(proxy): no-path-concat

julien-f committed:
chore(proxy/_listPartitions): use logger instead of console

julien-f committed:
feat(@xen-orchestra/proxy): 0.9.0

julien-f committed:
feat(proxy/TaskLogger): timestamp logs

julien-f committed:
fix(proxy/DeltaBackupWriter): delete old backups sequentially

julien-f committed:
fix(proxy/forkStreamUnpipe): destroy stream when no more consumers

julien-f committed:
feat(proxy/DeltaBackupWriter/checkBaseVdis): log errors as warnings

julien-f committed:
chore(proxy/VmBackup#_selectBase): variable casing

julien-f committed:
feat(@xen-orchestra/xapi): 0.4.1

julien-f committed:
feat(@xen-orchestra/proxy): 0.9.1

julien-f committed:
chore(proxy-cli): no-path-concat

julien-f committed:
fix(proxy): forbid local remotes

These remotes are local to the proxy appliance and not the XOA, this might be surprising.

Also, proxy appliances are supposed to be stateless which is not the case with local remotes.

We'll revert this in the future if a valid use case is presented.

These remotes are not disabled in dev mode because they are used for tests, but that might change in the future.

badrAZ committed:
fix(proxy/remote.test): always return a result object (#5485)

badrAZ committed:
fix(proxy/_SyncThenable.tryUnwrap): rejection handling (#5498)

julien-f committed:
Merge branch 'master' into proxy-v1

MathieuRA committed:
feat(proxy/vmBackup): support for backup modifier VM tags (#5507)

See #5450

badrAZ committed:
feat(proxy/backups): add backup size to delta backups' metadata (#5511)

julien-f committed:
feat(proxy): dynamic config via dedicated mixin (#5515)

badrAZ committed:
feat(proxy): ability to run metadata backup (#5484)

julien-f committed:
feat(proxy-cli): --file flag required

julien-f committed:
feat(proxy-cli): accept CSON files

Similar to JSON, CSON is based on CoffeeScript and is easier for humans:
- comments are supported
- keys don't need to be quoted
- arrays and objects don't need commas
- objects have no braces

julien-f committed:
chore(proxy-cli): replace return with await

julien-f committed:
feat(proxy-cli): file can contain multiple calls

julien-f committed:
chore(proxy-cli): process.stdout → stdout

julien-f committed:
feat(@xen-orchestra/proxy-cli): 0.2.0

badrAZ committed:
feat(proxy): ability to list metadata backups (#5505)

badrAZ committed:
feat(proxy/backups): ability to restore metadata backups (#5518)

badrAZ committed:
feat(proxy/backups): ability to delete metadata backups (#5521)

julien-f committed:
Merge branch 'master' into proxy-v1

julien-f committed:
feat(@xen-orchestra/proxy): 0.10.0

badrAZ committed:
feat(proxy/backups): move getTmpDir to its own module (#5552)

This makes `RemoteAdapter` independent of `Remotes`.

julien-f committed:
chore(proxy/_getTmpDir): own implementation

Does not depend on `tmp` anymore.

badrAZ committed:
feat(proxy/backups): make RemoteAdapter not depend on app hooks (#5554)

badrAZ committed:
feat(proxy/backups): make RemoteAdapter not depend on app (#5559)

julien-f committed:
feat(proxy/config): new method watchDuration

julien-f committed:
feat(proxy/debounceResource): dynamic config delay

badrAZ committed:
feat(proxy): move getAdapter from remotes to backups (#5563)

badrAZ committed:
fix(proxy/backups): fix resource function not called with backup context (#5566)

Introduced by c19326ed501629afdefe954a7ac375f9328aea63

badrAZ committed:
feat(proxy/backups): make Backup not depend on app (#5560)

badrAZ committed:
feat(proxy/backups): move fetchPartitionFiles implementation to RemoteAdapter (#5567)

badrAZ committed:
feat(proxy/xapi/vm): ability to ignore [NOBAK] VDIs on snapshot (#5543)

badrAZ committed:
feat(proxy/backups): move ImportVmBackup to its own module (#5569)

badrAZ committed:
feat(proxy/backups): move mount/unmount partition to its own module (#5570)

badrAZ committed:
feat(proxy/backups): log VM backup restoration (#5575)

julien-f committed:
feat(proxy/remotes): local remotes enablable via config

julien-f committed:
fix(xapi#VM_create): fix bios_string support

badrAZ committed:
feat(proxy/backups/RemoteAdapter): replace decorator syntax (#5583)

In an effort to make transpilation unnecessary when migrated to `@xen-orchestra/backups`.

badrAZ committed:
chore(proxy/backups/Backup): replace decorator syntax (#5586)

julien-f committed:
Merge branch 'master' into proxy-v1

julien-f committed:
Merge branch 'master' into proxy-v1

julien-f committed:
feat(proxy/config): add xapiOptions.ignoreNobakVdis setting

Introduced by d2256ce73e8a248b1718a12523c63ab1b26511c9

badrAZ committed:
chore(proxy): replace decorateResult by decorateWith(compose) (#5595)

badrAZ committed:
chore(@xen-orchestra/backups): mutualize modules (#5594)

badrAZ committed:
feat(proxy/app): use @vates/disposable/debounceResource (#5596)
